### DISCUSSION TITLE
What is the nuance in the no-inner-declarations documentation?

### DISCUSSION THREAD

**Discussion by ghost** [eslint/eslint#15229] [Q&A] [answered]
Hello, I'm new to ESLint and my code change complained about no-inner-declarations. I resolved it by assigning my function to a variable.

While reading the docs I became very confused. Can anyone explain the difference?

Incorrect:
https://github.com/eslint/eslint/blob/main/docs/rules/no-inner-declarations.md?plain=1#L79-L83
```js
function doSomethingElse() {
    if (test) {
        function doAnotherThing() { }
    }
}
```

Correct:
https://github.com/eslint/eslint/blob/main/docs/rules/no-inner-declarations.md?plain=1#L95-L97
```js
function doSomethingElse() {
    function doAnotherThing() { }
}
```

**Comment by nzakas** [comment]
In the first example, you are declaring a function inside of a non-function block (an if block). In the second, you are declaring a function directly inside a function. As the docs explain:

> In JavaScript, prior to ES6, a function declaration is only allowed in the first level of a program or the body of another function, though parsers sometimes [erroneously accept them elsewhere](https://code.google.com/p/esprima/issues/detail?id=422). This only applies to function declarations; named or anonymous function expressions can occur anywhere an expression is permitted.

Assuming you are in an ES6+ environment, you can safely disable this rule.

**Reply by ghost** [reply]
Oops - reposting as reply instead of answer:
Thank you @nzakas .
> a function declaration is only allowed in the first level of a program or the body of another function

This is the part that confuses me. In the incorrect example:
```
function doSomethingElse() {
    if (test) {
        function doAnotherThing() { }
    }
}
```
`function doAnotherThing()` _is_ declared within the body of another function.

Is this just a grammar error and it means:
>a function declaration is only allowed in the topmost scope of a program, or the topmost scope of a function body

**Comment by nzakas** [comment, marked as answer]
> function doAnotherThing() is declared within the body of another function.

Technically, yes, but the primary concern is the immediate parent of the function declaration.  If the immediate parent is not a program or function body, then the rule will warn.
